### DISCUSSION TITLE
Loading certificate file and key from inputstream or bytearray instead of file name

### DISCUSSION THREAD

**Discussion by lombaardcj007** [MatthiasValvekens/pyHanko#158] [Q&A]
Is it possible to call `SimpleSigner.load()` with a bytearray or string that represents the contents of the `key_file` and `cert_file`

[SimpleSigner load method](https://pyhanko.readthedocs.io/en/latest/api-docs/pyhanko.sign.signers.pdf_cms.html?highlight=SimpleSigner.load#pyhanko.sign.signers.pdf_cms.SimpleSigner.load)
`class method load(key_file, cert_file, ca_chain_files=None, key_passphrase=None, other_certs=None, signature_mechanism=None, prefer_pss=False)`

I want to keep the keys in a database instead of stored onto disk.

Sample code to sign PDF document:
```from pyhanko import stamp
from pyhanko.pdf_utils import text, images
from pyhanko.pdf_utils.font import opentype
from pyhanko.pdf_utils.incremental_writer import IncrementalPdfFileWriter
from pyhanko.sign import fields, signers

SELF_SIGN = signers.SimpleSigner.load(
    'cert/example.key',
    'cert/example.crt',
)

# define signature field
meta = signers.PdfSignatureMetadata(field_name='Signature')
pdf_signer = signers.PdfSigner(
    meta, signer=SELF_SIGN
)

with open('source/SCRIPT_unsigned.pdf', 'rb') as inf:
    w = IncrementalPdfFileWriter(inf)

    # generate output file
    with open('out/output.pdf', 'wb') as outf:
        pdf_signer.sign_pdf(w, output=outf)
```

**Comment by MatthiasValvekens** [comment]
Hi @lombaardcj007


Sure, that's possible, you just have to do what `SimpleSigner.load(...)` does, which is calling `SimpleSigner(...)` with the correct parameters ;).

All joking aside, precisely _what_ you have to do depends on how you intend to store the certificates and keys in your database, mainly whether you're storing the raw (binary) DER representation or its PEM-armored textual form.

If you're working from raw bytes, then e.g. a certificate can be loaded like this:

```python
from asn1crypto import x509


cert_der_bytes = fetch_from_database(...)

my_cert = x509.Certificate.load(cert_der_bytes)
```

If it's PEM-encoded, then you'll want to use something like this:

```
def _enum_certs(pem_bytes):
        if pem.detect(pem_bytes):
            pems = pem.unarmor(pem_bytes, multiple=True)   # allow more than 1 cert
            for type_name, _, der in pems:
                if type_name is None or type_name.lower() == 'certificate':
                    yield x509.Certificate.load(der)
```


Admittedly keys are a little more complicated, since they can be encoded in a variety of ways (depending on encryption settings and whatnot). If you're capable of retrieving a raw `PrivateKeyInfo` representation from your DB, then `PrivateKeyInfo.load(...)` is all you need. If you're using PKCS#8 as a storage format for your keys then you need an intermediate step.

...while I was typing this I planned to post a link to the convenience wrappers around `pyca/cryptography` in pyHanko that hide all this complexity from the user, but then I realised that those wrappers also expect file names. Silly me. 🤦‍♂️

Long story short, here's code that you can (more or less) trivially copy-paste and tweak to do what you want: https://github.com/MatthiasValvekens/pyHanko/blob/768959d3cb0d75dd10fb0abd01dfc38fb9af5cd9/pyhanko/sign/general.py#L393-L482.

In the next release, I'll provide a version that accepts an arbitrary binary IO handle.

**Reply by MatthiasValvekens** [reply]
Filed as #160.
